### PR TITLE
Remove matplotlib as a core dependency.

### PR DESCRIPTION
**Pull request by hmaarrfk**
Issue #2710

Move matplotlib to optional dependency.
Skip tests that can't run without matplotlib

Few questions:
  - `geometry.polygon_clip` requires `matplotlib path, transforms`. :
         - Look into updating the [celiagg](https://github.com/celiagg/celiagg) conda recipe so it isn't pinned to numpy 1.12. Ultimately, we decided against this since it seemed like we were just removing matplotliib and replacing it with a different dependency.
              - See this PR https://github.com/conda-forge/celiagg-feedstock/pull/4 on celiagg's feedstock
         - A python implemented of the required algorithm was written to replace the matplotlib exposed algorithm.

Note: This was rebased onto: https://github.com/scikit-image/scikit-image/pull/3432
## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

## References
[If this is a bug-fix or enhancement, it closes issue # ]
[If this is a new feature, it implements the following paper: ]

## For reviewers

(Don't remove the checklist below.)

- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.

### CODE REVIEW

**Comment by hmaarrfk** [comment]
@scikit-image/core I think this is mostly ready for discussion.

In a face to face discussion @stefanv had mentionned that he was OK with having the polygon clip functions not working without matplotlib. 

@jni on gitter asked if it was possible to vendor the code from matplotlib that would enable the use of the AGG functions. 

Unfrotunately, it seems that matplotlib [vendored AGG](https://github.com/matplotlib/matplotlib/tree/master/extern) themselves, and I don't feel like moving the compijlation burden to scikit-image

For reference, here are the problematic lines
https://github.com/scikit-image/scikit-image/blob/master/skimage/_shared/_geometry.py#L29

**Comment by jni** [comment]
@hmaarrfk eh, I'm happy to leave it as is (thanks!) and only worry about the performance if someone complains. I'd use `@` instead of `.dot` though. But I won't be a hardass about it. =P

Based on `git grep polygon_clip`, it looks like we only use polygon clipping in `skimage.draw.polygon_perimeter` and then only when `clip=True`, which is not the default. So my guess is that not many people if any depend on this functionality being fast. 

So for me this is :+1: conditional on tests passing. (Again if you want to remove the `.dot`s I won't stop you. ;)

File: requirements/docs.txt [hunk 1]
```diff
@@ -6,3 +6,4 @@ sphinx-gallery
 sphinx-copybutton
 pytest-runner
 scikit-learn
+matplotlib>=2.0.0
```
Comment (sciunto) [inline requirements/docs.txt]: Why dont we blacklist ,!=3.0.0 here as well?

Comment (hmaarrfk) [reply requirements/docs.txt]: you are right, I just wanted to hear back more general concerns about the methodology before cleaning this kind of stuff up. In fact, we definitely need to blacklist 3.0.0 for the docs.

**Comment by hmaarrfk** [comment]
@jni, fixed up the `@` you suggested.

**Comment by sciunto** [comment]
@soupault Are you available to give your feedback on this PR?

File: tools/travis/before_install.sh [hunk 2]
```diff
@@ -29,8 +29,12 @@ export WHEELHOUSE
 export DISPLAY=:99.0
 # This causes way too many internal warnings within python.
 # export PYTHONWARNINGS="d,all:::skimage"
-# codecov will combine different paths
-export TEST_ARGS="--doctest-modules --cov=skimage"
+if [[ "$OPTIONAL_DEPS" == "1" ]]; then
+  # We must run doctests with the optional deps since
+  # it seems to crawl through the code base. if it cannot find
+  # matplotlib, it causes an error when it gathers the tests
+  export TEST_ARGS="--doctest-modules --cov=skimage"
```
Comment (hmaarrfk) [inline tools/travis/before_install.sh]: Just bringing attention to this specific line. I really don't know how to deal with these doctests. @stefanv brought this up earlier in https://github.com/scikit-image/scikit-image/pull/3311#discussion_r231336994

Comment (hmaarrfk) [reply tools/travis/before_install.sh]: I had to add a test dependenecy on doctestplus https://github.com/astropy/pytest-doctestplus/

**Comment by hmaarrfk** [comment]
@stefanv suggests porting this code:
https://github.com/stefanv/supreme/blob/master/supreme/ext/polygon.c#L121

**Comment by soupault** [comment]
This clearly needs a rebase.

**Comment by hmaarrfk** [comment]
This is a difficult PR to keep updated as it touches build infrastructure.

The general idea of the PR is there, unless somebody else is willing to press merge as soon as the CIs pass, there is no point in spending the effort to rebase things.

@jni supported the general idea. 

If anybody wants to work together to get this through, I'm happy to spend the time rebasing/touching this up.

**Comment by soupault** [comment]
@hmaarrfk ok, sorry for the inconvenience. I'm going to make a thorough review in the coming days. Let's keep the PR in the current state.

**Comment by hmaarrfk** [comment]
No need to be sorry. I'm just stating that this PR will take more than a quick glance or quick rebase

**Comment by hmaarrfk** [comment]
@jni, opps. I saw a refressh button next to your review, and I decided to explore....

**Comment by jni** [comment]
> opps. I saw a refressh button next to your review, and I decided to explore....

😂

**Comment by jni** [comment]
Also I feel like those commits shouldn't be in this PR for some reason...?

**Comment by hmaarrfk** [comment]
yeah, sorry those commits have been moved to independent PRs. but this just speaks to the point I was trying to make earlier. I need the CIs to be working, to ensure that I don't break some crucial functionality in this PR....

File: requirements/docs.txt [hunk 3]
```diff
@@ -6,6 +6,7 @@ sphinx-gallery
 sphinx-copybutton
 pytest-runner
 scikit-learn
+matplotlib>=2.0.0,!=3.0.0
```
Comment (Borda) [inline requirements/docs.txt]: what about future >3.0.0 it still won't be compatible, right?

Comment (hmaarrfk) [reply requirements/docs.txt]: it should work, it is just 3.0.0 that is skipped. When 3.0.0 it had a bug, that was fixed in 3.0.1

Comment (Borda) [reply requirements/docs.txt]: well, now I see that there is no support for py2, (matplotlib 3.x droped it also)

Comment (hmaarrfk) [reply requirements/docs.txt]: for now, we still support installing matplotlib 2.3 + scikit-image 0.15

**Comment by alexandrejaguar** [comment]
Hey @hmaarrfk,
how are we on this one? :)

**Comment by hmaarrfk** [comment]
This needs a few core devs to want to take this on at one time. I can update if you want to review.

File: skimage/_shared/_geometry.py [hunk 4]
```diff
@@ -53,3 +61,81 @@ def polygon_area(pr, pc):
     pr = np.asarray(pr)
     pc = np.asarray(pc)
     return 0.5 * np.abs(np.sum((pc[:-1] * pr[1:]) - (pc[1:] * pr[:-1])))
+
+
+def _clip_sutherland_hodgman(rp, cp, rclip, cclip):
```
Comment (stefanv) [inline skimage/_shared/_geometry.py]: Where does this implementation come from? These clipping algorithms are notoriously tricky to get right.

Comment (hmaarrfk) [reply skimage/_shared/_geometry.py]: I wrote it adapting the code you referred to I believe.

https://github.com/stefanv/supreme/blob/master/supreme/ext/polygon.c#L121

**Comment by rfezzani** [comment]
Sad that this PR is not merged :confused:. @hmaarrfk do you have bandwidth to work again on it?

**Comment by hmaarrfk** [comment]
That is very true. If you are willing to help push this through, then I can likely mustard some courage to revive certain PRs.

**Comment by rfezzani** [comment]
@stefanv, even if `matplotlib` pedigree is well known, it's a huge dependency, and its risk analysis is far from being easy :sweat_smile:.

> then lazy load matplotlib as necessary (i.e., if you don't use it, it never gets pulled in).

Do you mean that it will be installed on user machine the first time a function involving it is called?

**Comment by jni** [comment]
Superseded by #5990. Thank you for getting the ball rolling 3+ years ago (!), @hmaarrfk!

**Pull request closed without merging**